sql/src/main/java/org/elasticsearch/action/bulk/IndicesCreatedObserver.java

```diff
@@ -0,0 +1,93 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package org.elasticsearch.action.bulk;
+
+import com.google.common.util.concurrent.FutureCallback;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateObserver;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.unit.TimeValue;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.TimeoutException;
+
+class IndicesCreatedObserver extends ClusterStateObserver {
+
+    private final IndicesCreatedPredicate predicate;
+
+    public IndicesCreatedObserver(ClusterService clusterService,
+                                  ESLogger logger,
+                                  Collection<String> indicesToWaitFor) {
+        super(clusterService, logger);
+        this.predicate = new IndicesCreatedPredicate(indicesToWaitFor);
+    }
+
+    public void waitForIndicesCreated(final FutureCallback<Void> callback) {
+        waitForNextChange(new Listener() {
+            @Override
+            public void onNewClusterState(ClusterState state) {
+                callback.onSuccess(null);
+            }
+
+            @Override
+            public void onClusterServiceClose() {
+                close();
+            }
+
+            @Override
+            public void onTimeout(TimeValue timeout) {
+                callback.onFailure(new TimeoutException("indices not created timed out after " + timeout.toString() ));
+            }
+        }, predicate);
+    }
+
+    /**
+     * changepredicate that waits until all given indices are available in
+     * the cluster state
+     */
+    private static class IndicesCreatedPredicate implements ChangePredicate {
+
+        private final Collection<String> indices;
+
+        IndicesCreatedPredicate(Collection<String> indices) {
+            this.indices = indices;
+        }
+
+        private boolean hasAllIndices(ClusterState clusterState) {
+            return Arrays.asList(clusterState.metaData().concreteAllOpenIndices()).containsAll(indices);
+        }
+
+        @Override
+        public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus, ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+            return newStatus == ClusterState.ClusterStateStatus.APPLIED && hasAllIndices(newState);
+        }
+
+        @Override
+        public boolean apply(ClusterChangedEvent changedEvent) {
+            return changedEvent.indicesCreated().containsAll(indices)
+                    || hasAllIndices(changedEvent.state());
```

Comment:
maybe subtract the createdIndices from the "todoIndices" and try to avoid the hasAllIndices call